governance/governance.go

```diff
@@ -374,39 +388,93 @@ func (w *GovernanceWorker) governAgreements() {
 						event_code = persistence.EC_CANCEL_AGREEMENT_NO_REPLYACK
 					}
 
-					eventlog.LogAgreementEvent(w.db, persistence.SEVERITY_INFO,
-						fmt.Sprintf("Start terminating agreement for %v. Reason: %v", ag.RunningWorkload.URL, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason)),
-						event_code,
-						ag)
-
-					w.cancelAgreement(ag.CurrentAgreementId, ag.AgreementProtocol, reason, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason))
-
-					// cleanup workloads
-					w.Messages() <- events.NewGovernanceWorkloadCancelationMessage(events.AGREEMENT_ENDED, events.AG_TERMINATED, ag.AgreementProtocol, ag.CurrentAgreementId, ag.GetDeploymentConfig())
-
-					// clean up microservice instances if needed
-					w.handleMicroserviceInstForAgEnded(ag.CurrentAgreementId, false)
+					w.cancelGovernedAgreement(&ag,
+							fmt.Sprintf("Start terminating agreement for %v. Reason: %v", ag.RunningWorkload.URL, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason)),
+							reason,
+							event_code)
 				}
 
 			} else {
-				// For finalized agreements, make sure the workload has been started in time
+				// For finalized agreements, make sure the workload has been started in time.
 				if ag.AgreementExecutionStartTime == 0 {
 					// workload not started yet and in an agreement ...
 					if (int64(ag.AgreementAcceptedTime) + (MAX_CONTRACT_PRELAUNCH_TIME_M * 60)) < time.Now().Unix() {
 						glog.Infof(logString(fmt.Sprintf("terminating agreement %v because it hasn't been launched in max allowed time. This could be because of a workload failure.", ag.CurrentAgreementId)))
 						reason := w.producerPH[ag.AgreementProtocol].GetTerminationCode(producer.TERM_REASON_NOT_EXECUTED_TIMEOUT)
-						w.cancelAgreement(ag.CurrentAgreementId, ag.AgreementProtocol, reason, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason))
-						// cleanup workloads if needed
-						w.Messages() <- events.NewGovernanceWorkloadCancelationMessage(events.AGREEMENT_ENDED, events.AG_TERMINATED, ag.AgreementProtocol, ag.CurrentAgreementId, ag.GetDeploymentConfig())
-						// clean up microservice instances if needed
-						w.handleMicroserviceInstForAgEnded(ag.CurrentAgreementId, false)
+						w.cancelGovernedAgreement(&ag,
+							fmt.Sprintf("Start terminating agreement for %v. Reason: %v", ag.RunningWorkload.URL, w.producerPH[ag.AgreementProtocol].GetTerminationReason(reason)),
+							reason,
+							persistence.EC_CANCEL_AGREEMENT_EXECUTION_TIMEOUT)
+					}
+				} else {
+					// Finalized agreements could become out of policy if the policy changes on the node. Verify that the existing agreement
```

Comment:
Can the following part be put in a function?

Reply:
yes, I agree. I think it needs to be part of the agreement protocol code, but since I know we will be revisting this we can clean it up then.